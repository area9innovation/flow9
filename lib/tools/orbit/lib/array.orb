// Array utility functions library for Orbit

// map: Apply a function to each element in an array
fn map(arr, f) = (
    fn mapHelper(arr, f, idx, result) = (
        if idx >= length(arr) then result
        else mapHelper(arr, f, idx + 1, result + [f(arr[idx])])
    );
    mapHelper(arr, f, 0, [])
);

// filter: Keep only elements that match a predicate
fn filter(arr, pred) = (
    fn filterHelper(arr, pred, idx, result) = (
        if idx >= length(arr) then result
        else if pred(arr[idx]) then 
            filterHelper(arr, pred, idx + 1, result + [arr[idx]])
        else 
            filterHelper(arr, pred, idx + 1, result)
    );
    filterHelper(arr, pred, 0, [])
);

// fold: Accumulate a result based on each array element and an accumulator
fn fold(arr, init, f) = (
    fn foldHelper(arr, acc, f, idx) = (
        if idx >= length(arr) then acc
        else foldHelper(arr, f(acc, arr[idx]), f, idx + 1)
    );
    foldHelper(arr, init, f, 0)
);

// mapi: Map with index
fn mapi(arr, f) = (
    fn mapiHelper(arr, f, idx, result) = (
        if idx >= length(arr) then result
        else mapiHelper(arr, f, idx + 1, result + [f(idx, arr[idx])])
    );
    mapiHelper(arr, f, 0, [])
);

// filteri: Filter with index
fn filteri(arr, pred) = (
    fn filteriHelper(arr, pred, idx, result) = (
        if idx >= length(arr) then result
        else if pred(idx, arr[idx]) then 
            filteriHelper(arr, pred, idx + 1, result + [arr[idx]])
        else 
            filteriHelper(arr, pred, idx + 1, result)
    );
    filteriHelper(arr, pred, 0, [])
);

// foldi: Fold with index
fn foldi(arr, init, f) = (
    fn foldiHelper(arr, acc, f, idx) = (
        if idx >= length(arr) then acc
        else foldiHelper(arr, f(idx, acc, arr[idx]), f, idx + 1)
    );
    foldiHelper(arr, init, f, 0)
);

// find: Find the first element that matches a predicate
fn find(arr, pred) = (
    fn findHelper(arr, pred, idx) = (
        if idx >= length(arr) then None()
        else if pred(arr[idx]) then Some(arr[idx])
        else findHelper(arr, pred, idx + 1)
    );
    findHelper(arr, pred, 0)
);

// findi: Find with index
fn findi(arr, pred) = (
    fn findiHelper(arr, pred, idx) = (
        if idx >= length(arr) then -1
        else if pred(arr[idx]) then idx
        else findiHelper(arr, pred, idx + 1)
    );
    findiHelper(arr, pred, 0)
);

// findDef: Find with default value
fn findDef(arr, pred, default) = (
    find(arr, pred) is (
        Some(value) => value;
        None() => default
    )
);

// tail: Return all elements except the first one
fn tail(arr) = (
    if length(arr) <= 1 then []
    else subrange(arr, 1, length(arr) - 1)
);

// tailFrom: Return all elements from a given index to the end
fn tailFrom(arr, idx) = (
    if idx >= length(arr) then []
    else subrange(arr, idx, length(arr) - idx)
);

// take: Return the first n elements of an array
fn take(arr, n) = (
    if n <= 0 then []
    else (
        fn takeHelper(arr, idx, n, result) = (
            if idx >= length(arr) || idx >= n then result
            else takeHelper(arr, idx + 1, n, result + [arr[idx]])
        );
        takeHelper(arr, 0, n, [])
    )
);

// iter: Call a function for each element (for side effects)
fn iter(arr, f) = (
    fn iterHelper(arr, f, idx) = (
        if idx >= length(arr) then 0 // Return 0 instead of ()
        else (
            f(arr[idx]);
            iterHelper(arr, f, idx + 1)
        )
    );
    iterHelper(arr, f, 0)
);

// iteri: Iterate with index
fn iteri(arr, f) = (
    fn iteriHelper(arr, f, idx) = (
        if idx >= length(arr) then 0 // Return 0 instead of ()
        else (
            f(idx, arr[idx]);
            iteriHelper(arr, f, idx + 1)
        )
    );
    iteriHelper(arr, f, 0)
);

// filtermap: Combine filter and map operations
fn filtermap(arr, f) = (
    fn filtermapHelper(arr, f, idx, result) = (
        if idx >= length(arr) then result
        else (
            let value = f(arr[idx]);
            value is (
                Some(val) => filtermapHelper(arr, f, idx + 1, result + [val]);
                None() => filtermapHelper(arr, f, idx + 1, result);
                _ => filtermapHelper(arr, f, idx + 1, result) // For safety
            )
        )
    );
    filtermapHelper(arr, f, 0, [])
);

// filtermapi: Filtermap with index
fn filtermapi(arr, f) = (
    fn filtermapHelper(arr, f, idx, result) = (
        if idx >= length(arr) then result
        else (
            let value = f(idx, arr[idx]);
            value is (
                Some(val) => filtermapHelper(arr, f, idx + 1, result + [val]);
                None() => filtermapHelper(arr, f, idx + 1, result);
                _ => filtermapHelper(arr, f, idx + 1, result) // For safety
            )
        )
    );
    filtermapHelper(arr, f, 0, [])
);

// removeIndex: Remove an element at a specific index
fn removeIndex(arr, idx) = (
    if idx < 0 || idx >= length(arr) then arr
    else (
        fn removeIndexHelper(arr, idx, currentIdx, result) = (
            if currentIdx >= length(arr) then result
            else if currentIdx = idx then removeIndexHelper(arr, idx, currentIdx + 1, result)
            else removeIndexHelper(arr, idx, currentIdx + 1, result + [arr[currentIdx]])
        );
        removeIndexHelper(arr, idx, 0, [])
    )
);

// removeRange: Remove a range of elements
fn removeRange(arr, start, count) = (
    if start < 0 || start >= length(arr) || count <= 0 then arr
    else (
        fn removeRangeHelper(arr, start, end, idx, result) = (
            if idx >= length(arr) then result
            else if idx >= start && idx < end then removeRangeHelper(arr, start, end, idx + 1, result)
            else removeRangeHelper(arr, start, end, idx + 1, result + [arr[idx]])
        );
        removeRangeHelper(arr, start, start + count, 0, [])
    )
);

// removeFirst: Remove the first occurrence of a specific element
fn removeFirst(arr, element) = (
    fn removeFirstHelper(arr, element, idx, removed, result) = (
        if idx >= length(arr) then result
        else if !removed && arr[idx] = element then 
            removeFirstHelper(arr, element, idx + 1, true, result)
        else 
            removeFirstHelper(arr, element, idx + 1, removed, result + [arr[idx]])
    );
    removeFirstHelper(arr, element, 0, false, [])
);

// removeAll: Remove all occurrences of a specific element
fn removeAll(arr, element) = (
    filter(arr, \x.x != element)
);

// contains: Check if an element exists in the array
fn contains(arr, element) = (
    fn containsHelper(arr, element, idx) = (
        if idx >= length(arr) then false
        else if arr[idx] = element then true
        else containsHelper(arr, element, idx + 1)
    );
    containsHelper(arr, element, 0)
);

// exists: Check if any element satisfies a predicate
fn exists(arr, pred) = (
    fn existsHelper(arr, pred, idx) = (
        if idx >= length(arr) then false
        else if pred(arr[idx]) then true
        else existsHelper(arr, pred, idx + 1)
    );
    existsHelper(arr, pred, 0)
);

// forall: Check if all elements satisfy a predicate
fn forall(arr, pred) = (
    fn forallHelper(arr, pred, idx) = (
        if idx >= length(arr) then true
        else if !pred(arr[idx]) then false
        else forallHelper(arr, pred, idx + 1)
    );
    forallHelper(arr, pred, 0)
);

// uniq: Remove duplicates from an array
fn uniq(arr) = (
    if length(arr) <= 1 then arr
    else (
        fn uniq_helper(arr, idx, result) = (
            if idx >= length(arr) then result
            else (
                if contains(result, arr[idx]) then
                    // Skip duplicate
                    uniq_helper(arr, idx + 1, result)
                else
                    // Keep this element
                    uniq_helper(arr, idx + 1, result + [arr[idx]])
            )
        );
        uniq_helper(arr, 0, [])
    )
);

// Push an element to an array
fn arrayPush(arr, elem) = (
	arr + [elem]
);