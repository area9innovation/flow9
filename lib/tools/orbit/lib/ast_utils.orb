// AST utility functions for Orbit

// Create a properly nested AST node by folding over an array of arguments
// This function handles arrays of any length, whereas makeAst expects exactly 2 arguments
fn makeFoldedAst(op, args) = (
    let len = length(args);
    
    if len = 0 then (
        // Handle empty array case - return an appropriate default value
        // This depends on the operator, but we'll return a neutral element
        if op = "&&" then true
        else if op = "||" then false
        else args  // Return empty array as fallback
    ) else if len = 1 then (
        // For single element, just return it without creating an operation
        args[0]
    ) else if len = 2 then (
        // For exactly two elements, we can use makeAst directly
        makeAst(op, args)
    ) else (
        // For 3+ elements, fold to create a proper binary tree
        // Start with the first element and fold over the rest
        let initialAst = args[0];
        let remainingArgs = subrange(args, 1, len - 1);
        
        // Fold over the remaining arguments, combining each with the accumulator
        fold(remainingArgs, initialAst, \acc, element -> (
            makeAst(op, [acc, element])
        ))
    )
);