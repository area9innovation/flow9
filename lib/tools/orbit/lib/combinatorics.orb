// Basic combinatorial functions
fn factorial(n) = (
	if n <= 1 then 1
	else n * factorial(n - 1)
);

fn binomial_coefficient(n, k) = (
	if k < 0 || k > n then 0
	else if k == 0 || k == n then 1
	else (
		// Use multiplicative formula to avoid large intermediate values
		fn binom_helper(i, result) = (
			if i > k then result
			else binom_helper(i + 1, result * (n - k + i) / i)
		);
		binom_helper(1, 1)
	)
);

fn permutations(n, r) = (
	if r < 0 || r > n then 0
	else if r == 0 then 1
	else (
		// Calculate n! / (n-r)!
		fold(range(n-r+1, n+1), 1, \product, i.product * i)
	)
);

// Generate permutations of an array
fn generate_permutations(arr) = (
	if length(arr) <= 1 then [arr]
	else (
		let result = [];
		fn perms_helper(i, perms) = (
			if i >= length(arr) then perms
			else (
				let element = arr[i];
				let rest = remove_at_index(arr, i);
				let sub_perms = generate_permutations(rest);

				let new_perms = fold(sub_perms, perms, \acc, p.acc + [[element] + p]);
				perms_helper(i + 1, new_perms)
			)
		);
		perms_helper(0, [])
	)
);

// Generate combinations of an array
fn generate_combinations(arr, r) = (
	if r > length(arr) then []
	else if r == 0 then [[]]
	else if r == length(arr) then [arr]
	else (
		if length(arr) == 0 then []
		else (
			let first = arr[0];
			let rest = subrange(arr, 1, length(arr) - 1);

			// Combinations that include the first element
			let with_first = map(generate_combinations(rest, r - 1),
													\comb.[first] + comb);

			// Combinations that exclude the first element
			let without_first = generate_combinations(rest, r);

			with_first + without_first
		)
	)
);
