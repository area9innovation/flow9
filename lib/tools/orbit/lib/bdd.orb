import lib/rewrite;
import lib/array;
import lib/logic;

/**
 * Binary Decision Diagrams (BDDs)
 * 
 * This library implements Binary Decision Diagrams (BDDs) for Boolean formulas,
 * based on Shannon's expansion theorem. BDDs provide a canonical representation
 * for Boolean functions, which makes equivalence checking efficient.
 * 
 * BDDs are represented using the if-then-else (Ite) constructor:
 * Ite(var, then_branch, else_branch)
 * 
 * Where:
 * - var is a decision variable
 * - then_branch is the BDD when var is true
 * - else_branch is the BDD when var is false
 * - Terminal nodes are true and false
 */

// ----- VARIABLE ORDERING -----

// Get a numeric value for a variable to determine ordering
fn var_order(var) = (
    if is_not(var) then var_order(get_not_arg(var))
    else var
);

// Check if an expression is a negation
fn is_not(expr) = (
    getConstructor(expr) == "!"
);

// Get the argument of a negation
fn get_not_arg(expr) = (
    expr is (!arg => arg; _ => expr)
);

// ----- BDD CONSTRUCTION -----

// Convert a Boolean formula to a BDD
fn to_bdd(expr) = (
    expr is (
        // Terminal cases
        true => true;
        false => false;
        
        // Negation
		¬e => negate_bdd(to_bdd(e));
        
        // Conjunction (AND)
        a ∧ b => apply_and(to_bdd(a), to_bdd(b));
        
        // Disjunction (OR)
		a ∨ b => apply_or(to_bdd(a), to_bdd(b));
        
        // Already a BDD - ensure consistency
        Ite(v, t, e) => make_ite(v, to_bdd(t), to_bdd(e));
        
        // Any other expression type
        // Variable becomes a simple BDD
        v => if is_var(v) then make_ite(v, true, false) else expr;
    )
);

// Check if an expression is a variable
fn is_var(expr) = (
    let c = getConstructor(expr);
    c != "Ite" && c != "!" && c != "&&" && c != "||" && 
    expr != true && expr != false
);

// Create an Ite node with reduction
fn make_ite(var, then_case, else_case) = (
    // Rule 1: If both branches are equal, eliminate the test
    if then_case == else_case then then_case
    else Ite(var, then_case, else_case)
);

// Negate a BDD
fn negate_bdd(bdd) = (
    bdd is (
        true => false;
        false => true;
        Ite(v, t, e) => make_ite(v, negate_bdd(t), negate_bdd(e));
        _ => !bdd  // For variables or unprocessed expressions
    )
);

// Apply AND operation to two BDDs
fn apply_and(bdd1, bdd2) = (
    // Terminal cases
    if bdd1 == false || bdd2 == false then false
    else if bdd1 == true then bdd2
    else if bdd2 == true then bdd1
    else bdd1 is (
        Ite(v1, t1, e1) => (
            bdd2 is (
                Ite(v2, t2, e2) => (
                    // Choose variable to expand based on ordering
                    let v_order1 = var_order(v1);
                    let v_order2 = var_order(v2);
                    if v_order1 == v_order2 then (
                        // Same variable - combine results
                        make_ite(v1, apply_and(t1, t2), apply_and(e1, e2))
                    ) else if v_order1 < v_order2 then (
                        // v1 comes first - expand on v1
                        make_ite(v1, apply_and(t1, bdd2), apply_and(e1, bdd2))
                    ) else (
                        // v2 comes first - expand on v2
                        make_ite(v2, apply_and(bdd1, t2), apply_and(bdd1, e2))
                    )
                );
                // bdd2 is not an Ite node
                _ => make_ite(v1, apply_and(t1, bdd2), apply_and(e1, bdd2))
            )
        );
        // bdd1 is not an Ite node
        _ => bdd2 is (
            Ite(v2, t2, e2) => make_ite(v2, apply_and(bdd1, t2), apply_and(bdd1, e2));
            _ => bdd1 && bdd2  // Neither is an Ite node
        )
    )
);

// Apply OR operation to two BDDs
fn apply_or(bdd1, bdd2) = (
    // Terminal cases
    if bdd1 == true || bdd2 == true then true
    else if bdd1 == false then bdd2
    else if bdd2 == false then bdd1
    else bdd1 is (
        Ite(v1, t1, e1) => (
            bdd2 is (
                Ite(v2, t2, e2) => (
                    // Choose variable to expand based on ordering
                    let v_order1 = var_order(v1);
                    let v_order2 = var_order(v2);
                    
                    if v_order1 == v_order2 then (
                        // Same variable - combine results
                        make_ite(v1, apply_or(t1, t2), apply_or(e1, e2))
                    ) else if v_order1 < v_order2 then (
                        // v1 comes first - expand on v1
                        make_ite(v1, apply_or(t1, bdd2), apply_or(e1, bdd2))
                    ) else (
                        // v2 comes first - expand on v2
                        make_ite(v2, apply_or(bdd1, t2), apply_or(bdd1, e2))
                    )
                );
                // bdd2 is not an Ite node
                _ => make_ite(v1, apply_or(t1, bdd2), apply_or(e1, bdd2))
            )
        );
        // bdd1 is not an Ite node
        _ => bdd2 is (
            Ite(v2, t2, e2) => make_ite(v2, apply_or(bdd1, t2), apply_or(bdd1, e2));
            _ => bdd1 || bdd2  // Neither is an Ite node
        )
    )
);

// ----- BDD TO LOGIC CONVERSION -----

fn makeOr(a, b) = (
	scatter(`||`([a,b]))
);

fn makeAnd(a, b) = (
	scatter(`&&`([a,b]))
);

fn makeNot(a) = (
	makeAst("!", ([a]))
);

// Convert a BDD back to a standard Boolean formula
fn from_bdd(bdd) = (
    bdd is (
        // Terminal cases
        true => true;
        false => false;
        
        // BDD node
        Ite(v, t, e) => (
            let then_expr = from_bdd(t);
            let else_expr = from_bdd(e);
            
            // Simplify common cases
            if then_expr == true && else_expr == false then v
            else if then_expr == false && else_expr == true then makeNot(v)
            else if then_expr == true then makeOr(v, else_expr)
            else if else_expr == false then makeAnd(v, then_expr)
			else if else_expr == true then makeOr(makeNot(v), then_expr)
            else if then_expr == false then makeAnd(makeNot(v), else_expr)
            else (
				makeOr(
					makeAnd(v, then_expr),
					makeAnd(makeNot(v), else_expr)
				)
			)
        );
        
        // Variable or other expression
        _ => bdd
    )
);

// ----- VARIABLE EXTRACTION -----

// Extract all variables from a logic expression in order
fn extract_variables(expr) = (
    expr is (
        // Terminal cases
        true => [];
        false => [];
        
        // Variable
        v => if is_var(v) then [v] else [];
        
        // Negation
        !e => extract_variables(e);
        
        // AND/OR - combine variables from both sides
        a && b => uniq(sort(extract_variables(a) + extract_variables(b)));
        a || b => uniq(sort(extract_variables(a) + extract_variables(b)));
        
        // BDD node
        Ite(v, t, e) => uniq(sort([v] + extract_variables(t) + extract_variables(e)));
        
        // Fallback
        _ => []
    )
);

// ----- UTILITY FUNCTIONS -----

// Evaluate a BDD with a variable assignment
fn evaluate_bdd(bdd, assignment) = (
    bdd is (
        // Terminal cases
        true => true;
        false => false;
        
        // BDD node - follow appropriate branch based on variable value
        Ite(v, t, e) => (
            let value = assignment(v);
            if value then evaluate_bdd(t, assignment) else evaluate_bdd(e, assignment)
        );
        
        // Variable or unprocessed expression
        _ => assignment(bdd)
    )
);

// ----- EQUIVALENCE CHECKING -----

// Check if two Boolean formulas are equivalent
fn are_equivalent(expr1, expr2) = (
    let bdd1 = to_bdd(expr1);
    let bdd2 = to_bdd(expr2);
    bdd1 == bdd2
);

// ----- QUANTIFICATION OPERATIONS -----

// Variable restriction: substitute variable with a value
fn bdd_restrict(node, var, value) = (
    // Terminal cases are unchanged by restriction
    if node == true || node == false then node
    else node is (
        Ite(v, t, e) => (
            if v == var then
                // Direct substitution when we find the variable
                if value then t else e
            else (
                // For other variables, recurse on both branches
                make_ite(v, bdd_restrict(t, var, value), bdd_restrict(e, var, value))
            )
        );
        // For variables or other expressions
        _ => if node == var then value else node
    )
);

// Existential quantification: ∃v.f
fn bdd_exists(var, node) = (
    // Compute f[v→0] ∨ f[v→1]
    apply_or(
        bdd_restrict(node, var, false),
        bdd_restrict(node, var, true)
    )
);

// Universal quantification: ∀v.f
fn bdd_forall(var, node) = (
    // Compute f[v→0] ∧ f[v→1]
    apply_and(
        bdd_restrict(node, var, false),
        bdd_restrict(node, var, true)
    )
);

// Apply existential quantification for multiple variables
fn bdd_exists_multiple(vars, node) = (
    if length(vars) == 0 then node
    else bdd_exists_multiple(
        tail(vars), 
        bdd_exists(head(vars), node)
    )
);

// Apply universal quantification for multiple variables
fn bdd_forall_multiple(vars, node) = (
    if length(vars) == 0 then node
    else bdd_forall_multiple(
        tail(vars), 
        bdd_forall(head(vars), node)
    )
);

// ----- MODEL COUNTING -----

// Calculate the number of satisfying assignments for a BDD
// Simple implementation without using cache
fn bdd_count_models(node, var_count) = (
    node is (
        // No satisfying assignments for false
        false => 0;
        
        // All remaining variable assignments satisfy true
        true => (
            // 2^remaining_vars
            let result = 1;
            fn calculate_power(base, exponent, acc) = (
                if exponent <= 0 then acc
                else calculate_power(base, exponent - 1, acc * base)
            );
            calculate_power(2, var_count, 1)
        );
        
        // For BDD nodes, compute recursively
        Ite(v, t, e) => (
            // Count satisfying assignments in both branches
            let low_count = bdd_count_models(e, var_count - 1);
            let high_count = bdd_count_models(t, var_count - 1);
            
            // Sum the counts
            low_count + high_count
        );
        
        // For variables or other expressions
        _ => if node then (
            // 2^var_count
            let result = 1;
            fn calculate_power(base, exponent, acc) = (
                if exponent <= 0 then acc
                else calculate_power(base, exponent - 1, acc * base)
            );
            calculate_power(2, var_count, 1)
        ) else 0
    )
);

// Find one satisfying assignment for a BDD
fn bdd_find_satisfying_assignment(node) = (
    if node == false then
        None()  // No satisfying assignment
    else if node == true then
        Some([])  // Empty assignment works
    else node is (
        Ite(v, t, e) => (
            // Try high branch first (v=true)
            let high_result = bdd_find_satisfying_assignment(t);
            high_result is (
                Some(assignment) => Some([Pair(v, true)] + assignment);
                None() => (
                    // Try low branch (v=false)
                    let low_result = bdd_find_satisfying_assignment(e);
                    low_result is (
                        Some(assignment) => Some([Pair(v, false)] + assignment);
                        None() => None()
                    )
                )
            )
        );
        // For variables, assign true
        _ => Some([Pair(node, true)])
    )
);