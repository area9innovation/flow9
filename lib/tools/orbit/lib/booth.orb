import lib/array;

// Linear time Booth's algorithm to find canonical form in cyclic groups
fn booth_canonical(arr) = (
	if length(arr) <= 1 then arr
	else (
		let n = length(arr);

		// Create an auxiliary array which is the original array concatenated with itself
		let double_arr = arr + arr;

		// Initialize failure function array with -1
		// (We use -1 instead of conventional 0 to simplify logic)
		let f = \i. if i < 0 || i >= n then -1 else f(i);

		// Compute the min rotation index in linear time
		fn compute_min_index() = (
			// Initialize starting values
			fn find_min(i, j, k) = (
				// We're done when we've checked all rotations
				if i + k >= 2 * n then j
				else if k >= n then find_min(i + 1, j, 0)
				else (
					// Compute the current indices we're comparing
					let i_idx = (i + k) % (2 * n);
					let j_idx = (j + k) % (2 * n);

					if double_arr[i_idx] < double_arr[j_idx] then
						// i rotation is smaller, j becomes i
						find_min(i, i, k + 1)
					else if double_arr[i_idx] > double_arr[j_idx] then
						// j rotation is smaller, i moves forward
						find_min(i + k + 1, j, 0)
					else
						// Equal so far, continue comparison
						find_min(i, j, k + 1)
				)
			);

			// Start with first two positions
			find_min(0, 0, 0)
		);

		let min_start = compute_min_index();

		// Build the canonical form starting from min_start
		fn build_result(idx, result) = (
			if idx >= n then result
			else build_result(idx + 1, result + [arr[(min_start + idx) % n]])
		);

		build_result(0, [])
	)
);

