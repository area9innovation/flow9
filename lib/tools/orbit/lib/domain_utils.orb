import lib/graph;

// Compare two domains in the hierarchy
// Returns: 
//   -1 if a is more specific than b (a ⊂ ... ⊂ b)
//    0 if a and b are the same
//    1 if b is more specific than a (b ⊂ ... ⊂ a)
//    None() if a and b are not directly related
fn compareDomains(graph, a, b) = (
    if (a == b) then 0  // Same domain
    else if (isReachable(graph, a, b)) then -1  // a is more specific than b
    else if (isReachable(graph, b, a)) then 1   // b is more specific than a
    else None()  // Not directly related
);

// Find the lowest domain from a list of domains
// The lowest domain is the most specific one in the hierarchy
fn findLowestDomain(graph, domains) = (
    if (length(domains) == 0) then None()
    else if (length(domains) == 1) then Some(domains[0])
    else (
        // Start with the first domain as our current lowest
        let initialLowest = domains[0];
        
        // Compare with each other domain to find the lowest
        fold(subrange(domains, 1, length(domains) - 1), initialLowest, \currentLowest, domain -> (
            compareDomains(graph, domain, currentLowest) is (
                -1 => domain;  // This domain is more specific than current lowest
                0 => currentLowest;  // Same domain, keep current
                1 => currentLowest;  // Current lowest is more specific
                None() => (
                    // Domains are not directly related, we keep the current lowest
                    // A more sophisticated approach might look at the full hierarchy
                    currentLowest
                )
            )
        ))
    )
);

// Get the most specific domain a node belongs to
// This is the main function to be used in the OGraph system
fn getNodeMostSpecificDomain(domainGraph, og, nodeId) = (
    // Get all domains the node belongs to
    let domains = getONodeBelongsTo(og, nodeId);
    
    // Find and return the lowest domain from the list
    findLowestDomain(domainGraph, domains)
);

"Domain Hierarchy Utility Library Successfully Loaded"