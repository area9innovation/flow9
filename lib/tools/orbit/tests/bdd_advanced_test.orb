import lib/bdd;

println("=== BDD Advanced Features: Quantification and Model Counting ===\n");

// Setup a test formula: (a && b) || (a && c)
// This represents: a && (b || c)
let test_formula = quote((a && b) || (a && c));
println("Test formula: " + prettyOrbit(test_formula));

// Convert to BDD
let test_bdd = to_bdd(test_formula);
println("\nAs BDD:");
println(prettyOrbit(test_bdd));

// Test variable restriction
println("\n--- Variable Restriction ---");

// Restrict a = true
let a_true = bdd_restrict(test_bdd, quote(a), true);
println("Formula with a=true:");
println(prettyOrbit(a_true));
println("Simplified: " + prettyOrbit(from_bdd(a_true)));
// Expected: b || c

// Restrict b = false
let b_false = bdd_restrict(test_bdd, quote(b), false);
println("\nFormula with b=false:");
println(prettyOrbit(b_false));
println("Simplified: " + prettyOrbit(from_bdd(b_false)));
// Expected: a && c

// Test quantification operations
println("\n--- Quantification Operations ---");

// Existential quantification
println("\n* Existential Quantification *");

// Exists b
let exists_b = bdd_exists(quote(b), test_bdd);
println("∃b.[(a && b) || (a && c)]:");
println(prettyOrbit(exists_b));
println("Simplified: " + prettyOrbit(from_bdd(exists_b)));
// Expected: a (because if a=true, we can always make the formula true by setting b=true)

// Exists a
let exists_a = bdd_exists(quote(a), test_bdd);
println("\n∃a.[(a && b) || (a && c)]:");
println(prettyOrbit(exists_a));
println("Simplified: " + prettyOrbit(from_bdd(exists_a)));
// Expected: b || c (because we need at least one of b or c to be true)

// Universal quantification
println("\n* Universal Quantification *");

// Forall c
let forall_c = bdd_forall(quote(c), test_bdd);
println("∀c.[(a && b) || (a && c)]:");
println(prettyOrbit(forall_c));
println("Simplified: " + prettyOrbit(from_bdd(forall_c)));
// Expected: a && b (because setting c=false requires a&&b to be true)

// Forall a
let forall_a = bdd_forall(quote(a), test_bdd);
println("\n∀a.[(a && b) || (a && c)]:");
println(prettyOrbit(forall_a));
println("Simplified: " + prettyOrbit(from_bdd(forall_a)));
// Expected: false (because setting a=false makes the whole formula false)

// Test model counting
println("\n--- Model Counting ---");

// Original formula (a && b) || (a && c) has 3 variables
let var_count = 3;  // a, b, c
let model_count = bdd_count_models(test_bdd, var_count);
println("Number of satisfying assignments for (a && b) || (a && c): " + i2s(model_count));
// Expected: 3 out of 8 possible assignments (a=1,b=1,c=0 / a=1,b=1,c=1 / a=1,b=0,c=1)

// Test a simple formula: a || b
let simple_formula = quote(a || b);
let simple_bdd = to_bdd(simple_formula);
let simple_count = bdd_count_models(simple_bdd, 2);
println("\nNumber of satisfying assignments for a || b: " + i2s(simple_count));
// Expected: 3 out of 4 possible assignments (a=1,b=0 / a=0,b=1 / a=1,b=1)

// Test another formula: !a && !b
let neg_formula = quote(!a && !b);
let neg_bdd = to_bdd(neg_formula);
let neg_count = bdd_count_models(neg_bdd, 2);
println("\nNumber of satisfying assignments for !a && !b: " + i2s(neg_count));
// Expected: 1 out of 4 possible assignments (a=0,b=0)

// Test finding a satisfying assignment
println("\n--- Finding Satisfying Assignment ---");

let sat_assignment = bdd_find_satisfying_assignment(test_bdd);
println("A satisfying assignment for (a && b) || (a && c): " + prettyOrbit(sat_assignment));
// Expected: Something like {a=true, b=true, c=...} or {a=true, b=..., c=true}

println("\n=== Tests Completed Successfully ===");