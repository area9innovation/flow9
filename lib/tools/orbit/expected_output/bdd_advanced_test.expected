=== BDD Advanced Features: Quantification and Model Counting ===
Test formula: ((a ∧ b) ∨ (a ∧ c))
As BDD:
Ite(a, Ite(b, true, Ite(c, true, false)), false)
--- Variable Restriction ---
Formula with a=true:
Ite(b, true, Ite(c, true, false))
Simplified: (b ∨ c)
Formula with b=false:
Ite(a, Ite(c, true, false), false)
Simplified: (a ∧ c)
--- Quantification Operations ---
* Existential Quantification *
∃b.[(a && b) || (a && c)]:
Ite(a, true, false)
Simplified: a
∃a.[(a && b) || (a && c)]:
Ite(b, true, Ite(c, true, false))
Simplified: (b ∨ c)
* Universal Quantification *
∀c.[(a && b) || (a && c)]:
Ite(a, Ite(b, true, false), false)
Simplified: (a ∧ b)
∀a.[(a && b) || (a && c)]:
false
Simplified: false
--- Model Counting ---
Number of satisfying assignments for (a && b) || (a && c): 3
Number of satisfying assignments for a || b: 3
Number of satisfying assignments for !a && !b: 1
--- Finding Satisfying Assignment ---
A satisfying assignment for (a && b) || (a && c): Some([Pair(a, true), Pair(b, true)])
=== Tests Completed Successfully ===
Result: 0
